straxen/plugins/peaklet_processing.py

```diff
@@ -601,47 +604,80 @@ def compute(self, peaklets):
+        list of the exclusive last index of peaklet to be merged
+        """
+        peaklet_gaps = peaklet_starts[1:] - peaklet_ends[:-1]
+        n_gaps = len(peaklet_gaps)
+
+        previous_valid_gap = np.arange(n_gaps) - 1
+        next_valid_gap = np.arange(n_gaps) + 1
+
+        # gaps_to_merge includes two imaginary gaps before and after all peaklets
+        gaps_to_merge = np.zeros(n_gaps + 2, np.bool_)
+
+        # Check gaps from small to large
+        for gap_i in np.argsort(peaklet_gaps):
+            left_i = previous_valid_gap[gap_i] + 1
+            # the right is inclusive
+            right_i = next_valid_gap[gap_i]
+
+            sum_area = areas[left_i:right_i+1].sum()
+
+            if peaklet_gaps[gap_i] > max_gap:
+                # From now on gaps are too large to merge
+                break
 
-            if cluster_stop - cluster_start > max_duration:
+            if sum_area > max_area:
+                # The merged peak would be too large
                 continue
 
-            # Recover left and right indices of the clusters.
-            # stops are inclusive for a few lines... sorry...
-            while peaklet_starts[left_i] < cluster_start:
-                left_i += 1
-            right_i = left_i
-            while peaklet_ends[right_i] < cluster_stop:
-                right_i += 1
-
-            if left_i == right_i:
-                # One peak, nothing to merge
+            if (peaklet_ends[right_i] - peaklet_starts[left_i]) > max_duration:
+                # The merged peak would be too long
                 continue
 
-            if types[left_i] != 2:
-                # Doesn't start with S2: do not merge
+            if peaklet_gaps[gap_i] > merge_s2_threshold(np.log10(sum_area), gap_thresholds):
+                # Check with varying threshold based on peak area after merging
                 continue
 
-            right_i += 1   # From here on, right_i is exclusive
+            # Mark the current gap merged and update info of its previous and next gap
+            # There's no previous gap for gap[0], and no next gap for gap[n_gaps-1]
+            if gap_i != 0:
+                next_valid_gap[previous_valid_gap[gap_i]] = next_valid_gap[gap_i]
+            if gap_i != n_gaps - 1:
+                previous_valid_gap[next_valid_gap[gap_i]] = previous_valid_gap[gap_i]
 
-            if areas[left_i:right_i].sum() > max_area:
-                continue
+            gaps_to_merge[gap_i + 1] = True
 
-            start_merge_at[n_to_merge] = left_i
-            end_merge_at[n_to_merge] = right_i
-            n_to_merge += 1
+        i_gaps_to_merge = np.where(gaps_to_merge)[0]
+        start_merge_at = i_gaps_to_merge[~gaps_to_merge[i_gaps_to_merge - 1]] - 1
+        # end_merge_at is now exclusive
+        end_merge_at = i_gaps_to_merge[~gaps_to_merge[i_gaps_to_merge + 1]] + 1
+
+        # Doesn't start with S2: do not merge
+        start_merge_with_s2 = types[start_merge_at] == 2
```

Comment:
I hope I understood this right. But, would it make sense to do this earlier, because now, if this is true, nothing is merged at all?

Reply:
Right, I don't see an obvious place to do this during the loop as the whole loop is by size so the time info is abandoned.
One thing we can do is loop over the non-s2 merge starts and find the first that's an S2.

Reply:
Yes, I also think that would be basically the only way to solve it if this is really a problem. Did you see any such worrisome topology in the data?

Reply:
Now I do see some strange topology that might be like this (starting with an unknown). This will increase the n_competing values around it, which is the other thing we need to be a little careful about with this update (or any clustering updates), see [the second strange topology example](https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenon1t:tzhu:straxenmerges2s#events_merged_split_lost_and_found)